browser/ui/views/brave_actions/brave_actions_container.cc

```diff
@@ -353,6 +353,15 @@ bool BraveActionsContainer::CanStartDragForView(View* sender,
 }
 // end ToolbarActionView::Delegate members
 
+// BraveActionsContainerDelegate members
+void BraveActionsContainer::OnPopupClosed(
+    const extensions::ExtensionId& extension_id) {
+  if (extension_id == brave_rewards_extension_id && rewards_service_) {
+    rewards_service_->OnRewardsPanelClosed(GetCurrentWebContents());
```

Comment:
It does seem odd to thread this event through the Rewards service (through rewards service observers, and then to the ads host object), when the rewards service doesn't even really know anything about UI panels.

I understand that there's no existing, good channel through which to send this event. It seems to me that the code that opens the popup (right now, by calling `ShowActionUI`), should be able to provide a means of getting notified when the UI is closed (whether via callback, delegate, etc.).

Would it be possible to add some kind of delegate support for `ShowActionUI` instead of adding `OnRewardsPanelClosed` to the Rewards service?

Reply:
Fixed, did the following (discussed in DM):
1) add BraveActionAPI::NotifyActionUIClosed method
2) call BraveActionAPI::NotifyActionUIClosed from BraveActionsContainer::OnPopupClosed
3) add new class BraveActionAPI::ActionUIObserver that has OnActionUIClosed() method
4) notify observers from BraveActionAPI::NotifyActionUIClosed method
5) inherit BraveAdsHost from BraveActionAPI::ActionUIObserver and start listening for OnActionUIClosed event